Password protect settings

elshimone opened the pull request:
# Title

Add panel to settings page to password protect settings.


## Pull Request Type
- [ ] Bugfix
- [X ] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
closes #2907 

## Description
Adds a panel to the settings page to password protect all settings.  The password is currently stored in the clear along with other persistent settings.
If no password is set, the panel prompts the user to set a password.  If a password is set, the other settings panels are hidden until the correct password is supplied.  If the password is supplied, the settings are made visible and an option to remove the password is enabled.

## Screenshots 

Dialog to set a password to prevent access to settings
![image](https://user-images.githubusercontent.com/698207/206284506-2ad8e3f9-2f31-4447-bf7a-c26a0b623c92.png)

Dialog to enter a password to access settings (other settings are hidden)
![image](https://user-images.githubusercontent.com/698207/206284598-0c2b3aa5-4135-4cc5-b564-d205c320ac00.png)

Incorrect password prompt
![image](https://user-images.githubusercontent.com/698207/206284677-4d8d49c8-2eb2-4e50-8765-13fc3727955c.png)

Dialog to lock or remove the password when correct password has been entered
![image](https://user-images.githubusercontent.com/698207/206284760-65351c9e-bf36-407d-9d05-8fda4aa76ed6.png)

## Testing

I have tested the PR manually by going through the following steps:
* Start Freetube
* Navigate to settings, observe settings are available
* Open password protect settings panel
* Enter a password, and click set
* Other settings panels are hidden, observe prompt to enter password to unlock
* Enter an incorrect password and click "unlock"; incorrect password dialog show.  Dismiss by clicking ok
* Enter correct password. Observe that settings are unlocked.
* Navigate away from setting, e.g. to subscriptions
* Return to settings, observe they are locked
* Restart app, observe settings are locked
* Enter correct password, click remove password button.  Observe settings are unlocked
* Restart app, observe settings are unlocked

I have not observed any regressions due to my changes to the ft-input component, but would like some advice on what I should look for here.

## Desktop
- **OS:**
Pop_OS
- **OS Version:**
22.04
- **FreeTube version:**
18.0.0
## Additional context

PikachuEXE commented:
I use `yarn dev` to start app
After setting password and restart via `yarn dev` again the settings is unlocked
Not sure if caused by `yarn dev` or what (But no issue for other settings so I dunno)

Also I am not sure if that section should be put on the top (as it might not be a frequently used setting)

PikachuEXE commented:
@ChunkyProgrammer I have just reviewed it lol

elshimone commented:
> I use `yarn dev` to start app After setting password and restart via `yarn dev` again the settings is unlocked Not sure if caused by `yarn dev` or what (But no issue for other settings so I dunno)
> 
> Also I am not sure if that section should be put on the top (as it might not be a frequently used setting)

Good point I will move it from the top to the bottom.  I'll look into the setting not being persisted, perhaps I broke something when I was refactoring the code.  I had previously seen my subscriptions and other settings being reset with yarn dev, but I have just checked it and that doesn't happen any more (but the password setting is lost).

elshimone commented:
Ok I have moved the password settings to the bottom and fixed the setting persistence issue.  I had not read the comments in settings.js properly, and was calling the autogenerated setter rather than the update call which actually persists things - it can't have ever worked before :blush:

PikachuEXE reviewed (dismissed):
Tested locally
Relying on others to code review settings related code

absidue reviewed (changes requested):
Thanks for your PR, the code works but I have to say I am not a fan of the UX, having the enter password box in a settings section that is collapsed by default, doesn't make for a great unlocking experience.

It might be better if you split this up into two parts, the set, change and clear password settings section and then move the unlocking part into the Settings page. You could use `v-if` and `v-else` and centre the unlock box in the middle of the screen.

elshimone commented:
Yes that is a good idea, I will change the UX as you suggest @absidue

elshimone commented:
I have split out the password entry:

![image](https://user-images.githubusercontent.com/698207/207665766-c87260f8-618b-4f6c-89d5-2a8971326622.png)

and password management section:

![image](https://user-images.githubusercontent.com/698207/207666080-a3c99d5a-5635-4d39-9d02-e05b5184ea5a.png)

 The password entry is focused so you can type the password straight in.

PikachuEXE commented:
Found an error:
![image](https://user-images.githubusercontent.com/1018543/207747506-44aad0f8-e090-4d4b-ba99-17e884992c9c.png)

elshimone commented:
Fixed.

absidue commented on src/renderer/components/password-dialog/password-dialog.js:
```diff
@@ -0,0 +1,66 @@
+import Vue from 'vue'
+import { mapActions } from 'vuex'
+import FtSettingsSection from '../ft-settings-section/ft-settings-section.vue'
+import FtSelect from '../ft-select/ft-select.vue'
+import FtCard from '../ft-card/ft-card.vue'
+import FtInput from '../ft-input/ft-input.vue'
+import FtToggleSwitch from '../ft-toggle-switch/ft-toggle-switch.vue'
+import FtFlexBox from '../ft-flex-box/ft-flex-box.vue'
+import FtButton from '../ft-button/ft-button.vue'
+import FtPrompt from '../ft-prompt/ft-prompt.vue'
+
+export default Vue.extend({
+  name: 'PasswordDialog',
+  components: {
+    'ft-settings-section': FtSettingsSection,
+    'ft-select': FtSelect,
+    'ft-input': FtInput,
+    'ft-card': FtCard,
+    'ft-toggle-switch': FtToggleSwitch,
+    'ft-flex-box': FtFlexBox,
+    'ft-button': FtButton,
+    'ft-prompt': FtPrompt,
```
Could you please remove the unused component imports, you only need to import and register the components that you use in the associated .vue file.

absidue commented on src/renderer/components/password-settings/password-settings.js:
```diff
@@ -0,0 +1,64 @@
+import Vue from 'vue'
+import { mapActions } from 'vuex'
+import FtSettingsSection from '../ft-settings-section/ft-settings-section.vue'
+import FtSelect from '../ft-select/ft-select.vue'
+import FtInput from '../ft-input/ft-input.vue'
+import FtToggleSwitch from '../ft-toggle-switch/ft-toggle-switch.vue'
+import FtFlexBox from '../ft-flex-box/ft-flex-box.vue'
+import FtButton from '../ft-button/ft-button.vue'
+import FtPrompt from '../ft-prompt/ft-prompt.vue'
+
+export default Vue.extend({
+  name: 'PasswordSettings',
+  components: {
+    'ft-settings-section': FtSettingsSection,
+    'ft-select': FtSelect,
+    'ft-input': FtInput,
+    'ft-toggle-switch': FtToggleSwitch,
+    'ft-flex-box': FtFlexBox,
+    'ft-button': FtButton,
+    'ft-prompt': FtPrompt,
```
Same for this file.

elshimone replied:
Sorry, that was a copy and paste error on my part.
